[cmake] fix set of APP_RENDER_SYSTEM on KodiConfig.cmake

AlwinEsch opened the pull request:
## Description
In the implementation of GLES under Windows found the foolish finding of OpenGL. This seems to exist systematically and makes it a little harder. Then found that "APP_RENDER_SYSTEM" of "KodiConfig.cmake" is not defined correctly (empty defined).

This request fixes this and uses the "platform" folder.
Parts of "Platform.cmake" are also included in PrepareEnv.cmake.

But good to include them also in `PrepareEnv.cmake` to fix for addons?

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented on cmake/scripts/common/PrepareEnv.cmake:
```diff
@@ -7,6 +7,18 @@ if(NOT DEFINED KODI_MIRROR)
+  file(GLOB _platformnames RELATIVE ${CORE_SOURCE_DIR}/cmake/platform/${CORE_SYSTEM_NAME}/
+                                    ${CORE_SOURCE_DIR}/cmake/platform/${CORE_SYSTEM_NAME}/*.cmake)
+  string(REPLACE ".cmake" " " _platformnames ${_platformnames})
+  message(FATAL_ERROR "invalid CORE_PLATFORM_NAME: ${CORE_PLATFORM_NAME_LC}\nValid platforms: ${_platformnames}")
```
With this c&p part from Platform.cmake I'm not sure, are still OS types defined in ./cmake/platform/ folder?

AlwinEsch commented:
How is the state about them, OK to bring in?

wsnipex commented on cmake/scripts/common/Platform.cmake:
```diff
@@ -1,6 +1,13 @@
+  set(PLATFORM_USED_SOURCE_DIR ${CORE_SOURCE_DIR})
+else()
+  set(PLATFORM_USED_SOURCE_DIR ${CMAKE_SOURCE_DIR})
+endif()
```
Please explain this block. Both kodi and addons have CORE_SOURCE_DIR defined and pointing to the same main kodi source root, so we should be able to use CORE_SOURCE_DIR directly

AlwinEsch replied:
Is a bit a hen and egg problem.

On Kodi build it must be included on begin and `CORE_SOURCE_DIR` not set:
https://github.com/xbmc/xbmc/blob/master/CMakeLists.txt#L32
This then becomes set by https://github.com/xbmc/xbmc/blob/master/CMakeLists.txt#L36, and I think https://github.com/xbmc/xbmc/blob/master/cmake/scripts/common/Macros.cmake#L8 need a bit the defines from Platform.cmake.

wsnipex replied:
from a quick look it shouldn't be a problem to set CORE_SOURCE_DIR on top of the main CMakeLists.txt

AlwinEsch replied:
Have checked it does not work in other way without a bigger rewrite.

- `Platform.cmake` define the `CORE_SYSTEM_NAME`
- `CORE_SYSTEM_NAME` required by `Macros.cmake`
- `CORE_SOURCE_DIR` on `common/Macros.cmake` not allowed to set before, otherwise `${CORE_SYSTEM_NAME}/Macros.cmake` is not included (where need `CORE_SYSTEM_NAME`)

Not sure if this request fix something on my screensavers.rsxs Jenkins problem, but maybe it bring something :-)

AlwinEsch replied:
@wsnipex what you think about?

wsnipex replied:
I still think this should be refactored, but since that needs deeper changes we can leave it as is for v18. But please add a TODO comment that this should be fixed for v19

AlwinEsch replied:
Also heard that v19 is planning to restructure the CMake for addons. Maybe we could integrate it with it (or maybe it's already ;-)). Otherwise, I will see what can be done.

AlwinEsch committed:
[cmake] fix set of APP_RENDER_SYSTEM on KodiConfig.cmake

In the implementation of GLES under Windows found the foolish finding of OpenGL. This seems to exist systematically and makes it a little harder. Then found that "APP_RENDER_SYSTEM" of "KodiConfig.cmake" is not defined correctly (empty defined).

This request fixes this and uses the "platform" folder.
Parts of "Platform.cmake" are also included in PrepareEnv.cmake.

TODO: Refactor on v19 to remove the if/else usage

Rechi commented:
@AlwinEsch This broke compiling binary addons via depends for GBM and Wayland platform.

AlwinEsch commented:
Can you give the error output, if not easy to fix this request becomes reverted from me.

Rechi commented:
```
CMake Error at /home/jenkins/workspace/LINUX-64-GBM/cmake/platform/linux/gbm.cmake:13 (message):
  You need to decide whether you want to use GL- or GLES-based rendering in
  combination with the GBM windowing system.  Please set GBM_RENDER_SYSTEM to
  either "gl" or "gles".  For normal desktop systems, you will usually want
  to use "gl".
Call Stack (most recent call first):
  /home/jenkins/workspace/LINUX-64-GBM/cmake/scripts/common/Platform.cmake:37 (include)
  /home/jenkins/workspace/LINUX-64-GBM/tools/depends/target/binary-addons/x86_64-linux-gnu-debug/build/depends/lib/kodi/PrepareEnv.cmake:3 (include)
  CMakeLists.txt:186 (include)
```
